src/libcore/iter/traits.rs

```diff
@@ -519,6 +519,31 @@ pub trait ExactSizeIterator: Iterator {
         assert_eq!(upper, Some(lower));
         lower
     }
+
+    ///
+    /// Returns whether the iterator is empty.
+    ///
+    /// This method has a default implementation using `self.len()`, so you
+    /// don't need to implement it yourself.
+    ///
+    /// # Examples
+    ///
+    /// Basic usage:
+    ///
+    /// ```
+    /// let mut one_element = [0].iter();
+    /// assert!(!one_element.is_empty());
+    ///
+    /// assert_eq!(one_element.next(), Some(0));
+    /// assert!(one_element.is_empty());
+    ///
+    /// assert_eq!(one_element.next(), None);
+    /// ```
+    #[inline]
+    #[unstable(feature = "exact_size_is_empty", issue = "0")]
+    fn is_empty(&self) -> bool {
+        self.len() == 0
```

Comment:
The `assert!` will be optimized out in most cases anyway; the `Iterator::size_hint` function of such iterators usually does something along the lines of `let len = ...; (len, Some(len))` which the compiler can see through and optimize out the `assert!` in `ExactSizeIterator::len`.